R/join_key.R

```diff
@@ -0,0 +1,52 @@
+#' Create a relationship between a pair of datasets
+#'
+#' @description `r lifecycle::badge("stable")`
+#'
+#' @details `join_key()` will create a relationship for the variables on a pair
+#' of datasets.
+#'
+#' @inheritParams mutate_join_keys
+#' @param dataset_2 (optional `character`) other dataset name. In case it is omitted, then it
+#' will create a primary key for `dataset_1`.
+#' @param keys (optionally named `character`) where `names(keys)` are columns in `dataset_1`
+#' with relationship to columns of `dataset_2` given by the elements in `keys`.
+#' If `names(keys)` is `NULL` then the same column names are used for both `dataset_1`
+#' and `dataset_2`.
+#'
+#' @return object of class `JoinKeySet` to be passed into `join_keys` function.
+#'
+#' @seealso [join_keys()]
+#'
+#' @export
+join_key <- function(dataset_1, dataset_2 = dataset_1, keys) {
+  checkmate::assert_string(dataset_1)
+  checkmate::assert_string(dataset_2)
+  checkmate::assert_character(keys, any.missing = FALSE)
+
+  if (length(keys) > 0) {
+    if (is.null(names(keys))) {
+      names(keys) <- keys
+    }
+
+    if (any(names(keys) == "")) {
+      names(keys)[names(keys) == "" & keys != ""] <- keys[names(keys) == "" & keys != ""]
+    }
+
+    stopifnot(!is.null(names(keys)))
+    stopifnot(!anyDuplicated(keys))
+    stopifnot(!anyDuplicated(names(keys)))
+  }
+
+  if (dataset_1 == dataset_2 && any(names(keys) != keys)) {
+    stop("Keys within a dataset must match exactly: keys = c('A' = 'B') are not allowed")
+  }
+
+  structure(
+    list(
+      dataset_1 = dataset_1,
+      dataset_2 = dataset_2,
+      keys = keys
+    ),
+    class = "JoinKeySet"
+  )
```

Comment:
What do you think if join_key produce following `list(dataset1 = list(dataset2 = keys))`. Then it would be aligned with `join_keys` structure

```r
jk <- join_keys(
  join_key("df1", "df1", "id"),
  join_key("df1", "df2", c("id" = "id2")),
  join_key("df2", "df2", "id2")
)
jk |> yaml::as.yaml() |> cat()
# df1:
#   df1: id
#   df2: id2
# df2:
#   df1: id
#   df2: id2

jk2 <- modifyList(jk, list(df1 = list(df2 = "ID2")))
jk2 |> yaml::as.yaml() |> cat()
# df1:
#   df1: id
#   df2: ID2
# df2:
#   df1: id
#   df2: id2
```

Reply:
I like it for that reason, but I dislike having to add some 3 getters functions (even if used as internal only) as it becomes harder to get the `join_key` attributes.

Given that there are only 7 occurrences in the code, I'll make the change

```R
dataset_1.join_key <- function(jk) {
  names(jk)
}

dataset_2.join_key <- function(jk) {
  names(jk[[1]])
}

key.join_key <- function(jk) {
  unname(jk[[1]][[1]])
}
```